presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PlanNodeDecorrelator.java

```diff
@@ -277,35 +276,40 @@ public PlanNodeDecorrelator(PlanNodeIdAllocator idAllocator, PlanVariableAllocat
 
         private Multimap<VariableReferenceExpression, VariableReferenceExpression> extractCorrelatedSymbolsMapping(List<Expression> correlatedConjuncts)
         {
-            // TODO: handle coercions and non-direct column references
             ImmutableMultimap.Builder<VariableReferenceExpression, VariableReferenceExpression> mapping = ImmutableMultimap.builder();
             for (Expression conjunct : correlatedConjuncts) {
+                // TODO: Handle non-direct column references for ArithmeticBinaryExpression, ArithmeticUnaryExpression, NotExpression, etc
                 if (!(conjunct instanceof ComparisonExpression)) {
                     continue;
                 }
 
                 ComparisonExpression comparison = (ComparisonExpression) conjunct;
-                if (!(comparison.getLeft() instanceof SymbolReference
-                        && comparison.getRight() instanceof SymbolReference
-                        && comparison.getOperator().equals(EQUAL))) {
-                    continue;
-                }
-
-                VariableReferenceExpression left = variableAllocator.toVariableReference(comparison.getLeft());
-                VariableReferenceExpression right = variableAllocator.toVariableReference(comparison.getRight());
-
-                if (correlation.contains(left) && !correlation.contains(right)) {
-                    mapping.put(left, right);
-                }
-
-                if (correlation.contains(right) && !correlation.contains(left)) {
-                    mapping.put(right, left);
+                // handle coercions and non-direct column references
+                if (comparison.getOperator().equals(EQUAL)) {
+                    List<VariableReferenceExpression> left = extractUniqueExpression(comparison.getLeft());
+                    List<VariableReferenceExpression> right = extractUniqueExpression(comparison.getRight());
+
+                    for (VariableReferenceExpression leftVariableExpression : left) {
+                        for (VariableReferenceExpression rightVariableExpression : right) {
+                            if (correlation.contains(leftVariableExpression) && !correlation.contains(rightVariableExpression)) {
```

Comment:
what does it mean now if there are multiple unique variables on each side of the equals. won't this map the left expression to all fo the right expressions (and similarly for the right code below). that seems incorrect. e.g. what happens if you have `t1.a = t2.a +t2.b`,or ` t1.a + t1.b = t2.a + t2.b`? or even `t1.a + t2.a = t1.b + t2.b`

Reply:
when there are multiple unique column variables on each side of the comparison expression, they need to be validated if each unique variable from each side is a correlated variable and can be mapped to the right side's unique variable. 

E.g. `t1.a = t2.a  + t2.b` is tranformed into 
`cast(t1.a) = t2.a+t2.b` which would have a  a correlation mapping variables of `t2.a` and `t2.b` and a `cast(t1.a)`. 
So now left variable `t1.a` is in a mappting to each of right side variables `t2.a` and `t2.b`.

For `t1.a + t2.a = t1.b + t2.b`, the correlated predicate is tranformed into 
`cast(t1.a) + t2.a = cast(t1.b) + t2.b` for which the correlation variables are `t2.a`, `t2.b` with `cast(t1.a)` and `cast(t1.b)`.
so the mapping becomes `t2.b`-> `t1.a1` and `t2.a`->`t1.b`
![image](https://user-images.githubusercontent.com/32091981/122984869-59a9a080-d352-11eb-937f-b55a0acc4dc7.png)

I also added UT test coverage on some cases mentioned here to resolve the confusion.